tap_nice_incontact/streams.py

```diff
@@ -604,6 +609,200 @@ def get_records(self,
+                raise Exception("data extraction job failure", job_id)
+
+            status = job_details['status']
+            LOGGER.info(f'Job status ({job_id}): {status}')
+
+            if status == 'SUCCEEDED':
+                ready = True
+            elif status == 'RUNNING':
+                time.sleep(poll_delay)
+            else:
+                # Job status can also be: Failed, Cancelled, and Expired.
+                # https://help.nice-incontact.com/content/recording/dataextractionapi.htm
+                raise ValueError("Data extraction job failure", job_details)
+
+        records = self.download_csv(job_details['result']['url'])
+        return records
+
+    def sync(self,
+             state: dict,
+             stream_schema: dict,
+             stream_metadata: dict,
+             config: dict,
+             transformer: Transformer) -> dict:
+        """
+        The sync logic for an incremental stream.
+
+        :param state: A dictionary representing singer state
+        :param stream_schema: A dictionary containing the stream schema
+        :param stream_metadata: A dictionnary containing stream metadata
+        :param config: A dictionary containing tap config data
+        :param transformer: A singer Transformer object
+        :return: State data in the form of a dictionary
+        """
+        
+
+        start_date = singer.get_bookmark(state,
+                                        self.tap_stream_id,
+                                        self.replication_key,
+                                        config['start_date'])
+        bookmark_datetime = singer.utils.strptime_to_utc(start_date)
+        max_datetime = bookmark_datetime        
+        
+        with metrics.record_counter(self.tap_stream_id) as counter:
+            for record in self.get_records(config, bookmark_datetime):
+                # Convert data keys for CSV outputs to the correct field format
+                record = convert_data_keys(record)
+
+                if self.convert_data_types:
+                    record = convert_data_types(record, stream_schema)
+
+                transformed_record = transformer.transform(record, stream_schema, stream_metadata)
+                # pylint: disable=line-too-long
+                record_datetime = singer.utils.strptime_to_utc(transformed_record[self.replication_key])
+                if record_datetime >= bookmark_datetime:
+                    singer.write_record(self.tap_stream_id, transformed_record)
+                    counter.increment()
+                    max_datetime = max(record_datetime, bookmark_datetime)
+
+            bookmark_date = singer.utils.strftime(max_datetime)
+
+        state = singer.write_bookmark(state,
+                                    self.tap_stream_id,
+                                    self.replication_key,
+                                    bookmark_date)
+        singer.write_state(state)
+        return state
```

Comment:
I'm sometimes not getting state messages which I think is because of reaching the backoff limit. These are my logs:
```
INFO Skipping stream: agents
INFO Skipping stream: contacts_completed
INFO Skipping stream: skills_summary
INFO Skipping stream: skills_sla_summary
INFO Skipping stream: teams_performance_total
INFO Skipping stream: wfm_skills_contacts
INFO Skipping stream: wfm_skills_dialer_contacts
INFO Skipping stream: wfm_skills_agent_performance
INFO Skipping stream: wfm_agents
INFO Skipping stream: wfm_agents_schedule_adherence
INFO Skipping stream: wfm_agents_scorecards
INFO Starting sync for stream: qm_workflows
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-04", "endDate": "2023-01-05", "version": "3"}
INFO Created job: c725ab84-3e18-4419-ab52-ef024aa7a2f8
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs/c725ab84-3e18-4419-ab52-ef024aa7a2f8 - Making GET request
INFO Job status (c725ab84-3e18-4419-ab52-ef024aa7a2f8): RUNNING
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs/c725ab84-3e18-4419-ab52-ef024aa7a2f8 - Making GET request
INFO Job status (c725ab84-3e18-4419-ab52-ef024aa7a2f8): SUCCEEDED
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-05", "endDate": "2023-01-06", "version": "3"}
INFO Backing off _make_request(...) for 15.6s (tap_nice_incontact.client.NiceInContact403Exception: API returned a 403 - Forbidden)
INFO Connection error detected for POST to /jobs, triggering backoff: 1 try
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-05", "endDate": "2023-01-06", "version": "3"}
INFO Backing off _make_request(...) for 15.0s (tap_nice_incontact.client.NiceInContact403Exception: API returned a 403 - Forbidden)
INFO Connection error detected for POST to /jobs, triggering backoff: 2 try
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-05", "endDate": "2023-01-06", "version": "3"}
INFO Created job: f0ee22ca-6e5b-4a0e-848b-2f9652ebc062
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs/f0ee22ca-6e5b-4a0e-848b-2f9652ebc062 - Making GET request
INFO Job status (f0ee22ca-6e5b-4a0e-848b-2f9652ebc062): RUNNING
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs/f0ee22ca-6e5b-4a0e-848b-2f9652ebc062 - Making GET request
INFO Job status (f0ee22ca-6e5b-4a0e-848b-2f9652ebc062): SUCCEEDED
INFO METRIC: {"type": "counter", "metric": "record_count", "value": 36, "tags": {"endpoint": "qm_workflows"}}
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-06", "endDate": "2023-01-07", "version": "3"}
INFO Backing off _make_request(...) for 15.1s (tap_nice_incontact.client.NiceInContact403Exception: API returned a 403 - Forbidden)
INFO Connection error detected for POST to /jobs, triggering backoff: 1 try
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-06", "endDate": "2023-01-07", "version": "3"}
INFO Backing off _make_request(...) for 15.6s (tap_nice_incontact.client.NiceInContact403Exception: API returned a 403 - Forbidden)
INFO Connection error detected for POST to /jobs, triggering backoff: 2 try
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-06", "endDate": "2023-01-07", "version": "3"}
INFO Backing off _make_request(...) for 15.0s (tap_nice_incontact.client.NiceInContact403Exception: API returned a 403 - Forbidden)
INFO Connection error detected for POST to /jobs, triggering backoff: 3 try
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-06", "endDate": "2023-01-07", "version": "3"}
INFO Backing off _make_request(...) for 15.6s (tap_nice_incontact.client.NiceInContact403Exception: API returned a 403 - Forbidden)
INFO Connection error detected for POST to /jobs, triggering backoff: 4 try
INFO https://na1.nice-incontact.com/data-extraction/v1/jobs - Making POST request with data {"entityName": "qm-workflows", "startDate": "2023-01-06", "endDate": "2023-01-07", "version": "3"}
ERROR Giving up _make_request(...) after 5 tries (tap_nice_incontact.client.NiceInContact403Exception: API returned a 403 - Forbidden)
INFO METRIC: {"type": "counter", "metric": "record_count", "value": 60, "tags": {"endpoint": "qm_workflows"}}
```

Reply:
Will reach out on slack, can you send me your config/catalog? When I run it, I'm seeing it successfully emit the state even if it aborts making requests
```
ERROR Giving up _make_request(...) after 5 tries (tap_nice_incontact.client.NiceInContact403Exception: API returned a 403 - Forbidden)
INFO METRIC: {"type": "counter", "metric": "record_count", "value": 61, "tags": {"endpoint": "qm_workflows"}}
{"type": "STATE", "value": {"currently_syncing": "qm_workflows", "bookmarks": {"qm_workflows": {"jobEndDate": "2022-12-02T00:00:00.000000Z"}}}}
{"type": "STATE", "value": {"currently_syncing": "qm_workflows", "bookmarks": {"qm_workflows": {"jobEndDate": "2022-12-02T00:00:00.000000Z"}}}}
{"type": "STATE", "value": {"currently_syncing": null, "bookmarks": {"qm_workflows": {"jobEndDate": "2022-12-02T00:00:00.000000Z"}}}}
```